[Blaze] Payment methods list screen

selanthiraiyan opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #11717 
<!-- Id number of the GitHub issue this PR addresses. -->

## Description

Adds the screen that lists the saved payment methods from where the user can select a payment method for the Blaze campaign creation flow. Also, the user can add a new payment method from this screen.

Apologies for the 500+ lines of changes. They are due to unit tests. 

#### Changes
- Add a new view and view model for listing Blaze payment methods
- Change the `BlazeConfirmPaymentView` and its view model to allow payment selection. 
- Unit tests

#### Note

The adding payment method feature hasn't been tested yet. I added a separate subtask to test adding a payment method. We can pick up this work this once the API is ready. 

## Testing instructions
- Log in to a store that's eligible for Blaze.
- Navigate to the Products tab, select a published product and tap Promote with Blaze.
- On the campaign creation form, tap Confirm Details.
- Tap on the card info row to open the payment method list screen
- Verify that the WPCOM account name, email are displayed correctly in the list footer.
- Change the payment method, tap `Save`, and verify that the selection is saved.

## Screenshots

#### No payment methods available
| Dark | Light |
|--------|--------|
| <img width="320" alt="image" src="https://github.com/woocommerce/woocommerce-ios/assets/524475/ce3f05a2-6776-4d95-9545-72f885816003"> | <img width="323" alt="image" src="https://github.com/woocommerce/woocommerce-ios/assets/524475/575dbf97-076f-41a6-a51d-89a83dcf40b0"> | 

#### Payment methods available
| Dark | Light |
|--------|--------|
| <img width="324" alt="image" src="https://github.com/woocommerce/woocommerce-ios/assets/524475/4b2b9030-4b9c-436e-951d-95ce64f90e2c"> | <img width="327" alt="image" src="https://github.com/woocommerce/woocommerce-ios/assets/524475/072b1580-2abc-404d-8bd2-34d5ee16d56d"> | 

![Simulator Screen Recording - iPhone 15 Pro - 2024-01-18 at 20 37 53](https://github.com/woocommerce/woocommerce-ios/assets/524475/6b101821-2642-478f-9b32-50fca50a5a65)

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

selanthiraiyan committed:
Stub payment method response.

selanthiraiyan committed:
Make `BlazePaymentMethod` confirm to `Identifiable` protocol.

selanthiraiyan committed:
Add view model Blaze payment methods view.

selanthiraiyan committed:
Add unit tests for `BlazePaymentMethodsViewModel`.

selanthiraiyan committed:
Remove unused storage manager.

selanthiraiyan committed:
Add view for Blaze payments list screen.

selanthiraiyan committed:
Present payment method screen and update selection.

selanthiraiyan committed:
Ability to inject email and username.

selanthiraiyan committed:
Add more cards in stubbed response.

selanthiraiyan committed:
Add secure transactions header view.

dangermattic commented:
<!--
  0 Errors
  2 Warnings: This PR is larger than 500 lin...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 500 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-ios/milestone/181">17.0</a>. The due date for this milestone has already passed.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

selanthiraiyan committed:
Fix localization comment.

selanthiraiyan committed:
Log error if payment info not available.

selanthiraiyan committed:
Show empty payments mode in preview.

selanthiraiyan committed:
Remove comment.

selanthiraiyan committed:
Stub 3 cards in temporary response.

selanthiraiyan committed:
Merge branch 'trunk' into feat/11717-payment-list

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Blaze/ConfirmPayment/BlazeConfirmPaymentView.swift:
```diff
@@ -79,6 +79,9 @@ struct BlazeConfirmPaymentView: View {
                 .interactiveDismissDisabled()
         }
+        .sheet(isPresented: $viewModel.showAddPaymentSheet) {
+            BlazePaymentMethodsView(viewModel: viewModel.paymentMethodsViewModel!)
```
Instead of forced unwrapping, you can use `map` to unwrap the optional value:

```suggestion
            viewModel.paymentMethodsViewModel.map {
                BlazePaymentMethodsView(viewModel: $0)
            }
```

selanthiraiyan replied:
Updated in 1f32430929834523c8cb80f3f8bcf8a83d8a6c78

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Blaze/ConfirmPayment/BlazeConfirmPaymentView.swift:
```diff
@@ -145,7 +148,7 @@ private extension BlazeConfirmPaymentView {
     var addPaymentMethodButton: some View {
         Button {
-            // TODO: open payment method list
+            viewModel.showAddPaymentSheet = true
```
🤔 Now that I think of it, maybe we should show the web view directly from here to add a new method rather than asking users to tap that button again in the method list. WDYT?

selanthiraiyan replied:
Good point. Added this as a subtask in #11737

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Blaze/ConfirmPayment/BlazeConfirmPaymentViewModel.swift:
```diff
@@ -10,11 +10,26 @@ final class BlazeConfirmPaymentViewModel: ObservableObject {
     private let completionHandler: () -> Void
 
     private(set) var selectedPaymentMethod: BlazePaymentMethod?
+    @Published var showAddPaymentSheet: Bool = false
```
Nit: since this is not used in the view model, should we move it to the view as a private `@State` variable?

selanthiraiyan replied:
Done in ea7cad230c1ccc2fc0126cb6c79cb80526d9a700

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Blaze/ConfirmPayment/BlazeConfirmPaymentViewModel.swift:
```diff
@@ -97,4 +108,16 @@ private extension BlazeConfirmPaymentViewModel {
+    func setSelectedPaymentMethod(id paymentMethodID: String) {
+        guard let paymentMethod = paymentInfo?.savedPaymentMethods.first(where: { $0.id == paymentMethodID }) else {
+            return
+        }
```
When a new payment method is added, the existing method list will be unable to find the method by the ID. In that case, we should fetch the list again before trying to select the new one.

selanthiraiyan replied:
Great catch. Done in e8943656b034368d96bf15474d033319792c9f32. I will test and update this again when I work on #11737

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Blaze/ConfirmPayment/BlazeConfirmPaymentViewModel.swift:
```diff
@@ -97,4 +108,16 @@ private extension BlazeConfirmPaymentViewModel {
+            return
+        }
+        selectedPaymentMethod = paymentMethod
+        let rawCardType = paymentMethod.info.type
+        let cardType = CreditCardType(rawType: rawCardType)
+        cardIcon = cardType.icon
+        cardTypeName = paymentMethod.info.type
+        cardName = paymentMethod.name
```
Nit: should we move this to a separate method and trigger it whenever `selectedPaymentMethod` is updated (on `didSet`)?

selanthiraiyan replied:
Done in 2113f89e8940e70fba7f736597bf5ee76db08d7e and 831ce6987031af7d480d75f5344cc2cdf12914e5

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Blaze/ConfirmPayment/BlazePaymentMethodsView.swift:
```diff
@@ -0,0 +1,276 @@
+                            .resizable()
+                            .aspectRatio(contentMode: .fit)
+                            .frame(width: scale * Layout.SecureHeader.iconWidth, height: scale * Layout.SecureHeader.iconHeight)
+                            .tint(Color(.accent))
```
`tint` doesn't work for me, please use `foregroundColor` instead.

selanthiraiyan replied:
Thanks. My eyes didn't catch this. Also changed in 2c4e6ba1576f3db9cd1ab5a5370be1113acd1637

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Blaze/ConfirmPayment/BlazePaymentMethodsView.swift:
```diff
@@ -0,0 +1,276 @@
+                    // Empty state when there are no payments methods
+                    VStack(alignment: .center) {
+                        Spacer()
+                            .frame(height: Layout.spacerHeight)
```
Nit: Instead of adding this Spacer, you can set a top padding to the `EmptyState` view below.

selanthiraiyan replied:
Also changed in 2c4e6ba1576f3db9cd1ab5a5370be1113acd1637

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Blaze/ConfirmPayment/BlazePaymentMethodsView.swift:
```diff
@@ -0,0 +1,276 @@
+                        }()
+
+                        SelectableItemRow(title: "\(method.type.rawValue.capitalized) ****\(method.info.lastDigits)",
+                                          subtitle: method.name,
```
According to the design, we should display the card type and cardholder name, not the method type and name.

```suggestion
                        SelectableItemRow(title: "\(method.info.type) ****\(method.info.lastDigits)",
                                          subtitle: method.info.cardholderName,
```

selanthiraiyan replied:
Changed this as part of the refactor to list work in 2c4e6ba1576f3db9cd1ab5a5370be1113acd1637

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Blaze/ConfirmPayment/BlazePaymentMethodsView.swift:
```diff
@@ -0,0 +1,276 @@
+                    }
+                }
+
+                ToolbarItem(placement: .confirmationAction) {
+                    Button(action: {
+                        viewModel.saveSelection()
+                        dismiss()
+                    }, label: {
+                        if viewModel.isFetchingPaymentInfo {
+                            ProgressView()
+                        } else {
+                            Text(Localization.doneButton)
+                        }
+                    })
+                    .disabled(!viewModel.isDoneButtonEnabled)
+                }
```
👀  I don't see the Done button on the design. Do you think it'd be a nicer UX if we dismiss the screen immediately when a user selects a method or successfully adds a new method?

selanthiraiyan replied:
The current web view implementation (from the shipping label feature) doesn't send back the newly added payment ID. 

The Blaze API P2 states the following, peeHDf-2ii-p2

> If user selects to add a new payment method client should open a webview with the URL provided in add_payment_method.form_url and monitor webview URL changes and when URL matches add_payment_method.success_url, client should get the added payment method ID from the success URL GET parameter named add_payment_method.id_url_parameter, close the webview and submit the campaign with the obtained ID.

I added a subtask to check this in #11737

itsmeichigo replied:
Yeah, my understanding is that you check for the success URL provided in `BlazeAddPaymentInfo` and check for the param to get the ID. But good point to handle that in a separate PR.

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Blaze/ConfirmPayment/BlazePaymentMethodsView.swift:
```diff
@@ -0,0 +1,276 @@
+                        await viewModel.syncPaymentInfo()
+                    }
+                    let notice = Notice(title: Localization.paymentMethodAddedNotice, feedbackType: .success)
+                    ServiceLocator.noticePresenter.enqueue(notice: notice)
```
We should switch to using the `notice` modifier on the SwiftUI view instead of this for a more reliable way of presenting.

selanthiraiyan replied:
Done in 34d55e0d87ae46e105caa99b4ba12ce4c60a71dc

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Blaze/ConfirmPayment/BlazePaymentMethodsView.swift:
```diff
@@ -0,0 +1,276 @@
+                    // Payment Methods list
+                    ListHeaderView(text: Localization.paymentMethodsHeader, alignment: .left)
+                        .textCase(.uppercase)
+                        .renderedIf(viewModel.paymentMethods.isNotEmpty)
```
Suggestion: Instead of using the custom views, should we use a `List` with header and footer for simplicity? I think it makes sense to keep the "All transactions are secured..." header and button fixed.

```Swift
VStack {
    HStack { // header here }
    List(viewModel.paymentMethods) { 
        // content
    } header: {
        // TODO
    } footer: {
       // TODO
    } 
}
.safeAreaInset(edge: .bottom) {
    // Add new method button
}
```

selanthiraiyan replied:
Thanks for the suggestion. Done in 2c4e6ba1576f3db9cd1ab5a5370be1113acd1637

itsmeichigo commented on WooCommerce/WooCommerceTests/ViewRelated/Blaze/BlazePaymentMethodsViewModelTests.swift:
```diff
@@ -0,0 +1,245 @@
+                                                     stores: stores) { _ in }
+
+        // Then
+        XCTAssertEqual(viewModel.addPaymentMethodURL, URL(string: "https://example.com/blaze-pm-add")!)
```
```suggestion
        XCTAssertEqual(viewModel.addPaymentMethodURL, try XCTUnwrap(URL(string: "https://example.com/blaze-pm-add")))
```

selanthiraiyan replied:
Done in 86096c72e25fe1ed84f6224abd897e6e5b5cf08a

itsmeichigo reviewed (commented):
Hiya - I see a few differences between the screen with the design:

<img width="708" alt="Screenshot 2024-01-19 at 09 59 32" src="https://github.com/woocommerce/woocommerce-ios/assets/5533851/5623bdfb-54f7-44f3-9529-7335165eaee6">

I left some comments below for suggestions.

selanthiraiyan committed:
Refactor to use List.

selanthiraiyan committed:
Move `showAddPaymentSheet` from view model to view.

selanthiraiyan committed:
Stop force unwrapping view model.

selanthiraiyan committed:
Reload payment methods before attempting to select newly added payment method.

selanthiraiyan committed:
Move mock waiting time to unit tests.

selanthiraiyan committed:
Hard code card numbers to always return same stubbed response.

selanthiraiyan committed:
Add helper to display selected payment.

selanthiraiyan committed:
Remove helper method.

selanthiraiyan committed:
Use notice modifier.

selanthiraiyan committed:
Use XCTUnwrap instead of force unwrapping.

selanthiraiyan committed:
Merge branch 'trunk' into feat/11717-payment-list

selanthiraiyan committed:
Use `where` condition to get first matching `ProductVariationAction`.

selanthiraiyan commented:
Thanks for the review and suggestions, Huong! 

I refactored the view to use `List`. 
| Dark | Light |
|--------|--------|
| <img width="323" alt="image" src="https://github.com/woocommerce/woocommerce-ios/assets/524475/da15868c-1966-4b79-b7fc-98101ddb357c"> | <img width="322" alt="image" src="https://github.com/woocommerce/woocommerce-ios/assets/524475/4464878d-0104-4637-801c-ad4591ee0f05"> | 

I have responded to your comments about web view and added relevant subtasks in https://github.com/woocommerce/woocommerce-ios/issues/11737

The PR is ready for another look. 🙇

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Blaze/ConfirmPayment/BlazePaymentMethodsView.swift:
```diff
@@ -0,0 +1,303 @@
+                    Text(buttonText)
+                }
+                .buttonStyle(SecondaryButtonStyle())
+                .padding(Layout.ctaPadding)
```
We might need to add a background color to avoid showing list content below the footer view.

selanthiraiyan replied:
Added background color in https://github.com/woocommerce/woocommerce-ios/pull/11728/commits/9c244da1243fe3bb6dbaf889eb2522e9f8d2fa89. 

<img src="https://github.com/woocommerce/woocommerce-ios/assets/524475/edb47048-d790-4d2a-a5eb-ba64a536f1ae" width="320"/>

itsmeichigo reviewed (approved):
Thank you for the updates! 

I found an issue when logging in with application password to a site without Jetpack. The details for payment method list cannot fetch the WPCom account information as expected:

<img src="https://github.com/woocommerce/woocommerce-ios/assets/5533851/0ffb44cd-c820-46e7-9b3b-6f6192da4af6" width=320 />

I wonder if we should require WPCom login before directing to the campaign creation flow for users logged in without WPCom? That's what I see how they handle the case on the web.

I'm approving this PR regardless - please feel free to add another subtask to check this case.

selanthiraiyan committed:
Add background color to CTA section.

selanthiraiyan commented:
> I wonder if we should require WPCom login before directing to the campaign creation flow for users logged in without WPCom? That's what I see how they handle the case on the web.

Added a subtask in #11512
